### PR TITLE
bpo-23819: asyncio: Replace AssertionError with proper excs

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue23819

### CODE REVIEW

File: Lib/asyncio/events.py [hunk 1]
```diff
@@ -661,7 +662,8 @@ def get_event_loop(self):
     def set_event_loop(self, loop):
         """Set the event loop."""
         self._local._set_called = True
-        assert loop is None or isinstance(loop, AbstractEventLoop)
+        if loop is not None and not isinstance(loop, AbstractEventLoop):
+            raise TypeError(f'loop must be an instance of AbstractEventLoop or None, not {type(loop)}')
```
Comment (AlexWaygood) [inline Lib/asyncio/events.py]:
```suggestion
            raise TypeError(f"loop must be an instance of AbstractEventLoop or None, not '{type(loop)}'")
```

File: Lib/asyncio/events.py [hunk 2]
```diff
@@ -745,7 +747,8 @@ def set_event_loop_policy(policy):
 
     If policy is None, the default policy is restored."""
     global _event_loop_policy
-    assert policy is None or isinstance(policy, AbstractEventLoopPolicy)
+    if policy is not None and not isinstance(policy, AbstractEventLoopPolicy):
+        raise TypeError(f'policy must be an instance of AbstractEventLoopPolicy or None, not {type(policy)}')
```
Comment (AlexWaygood) [inline Lib/asyncio/events.py]:
```suggestion
        raise TypeError(f"policy must be an instance of AbstractEventLoopPolicy or None, not '{type(policy)}'")
```

File: Lib/asyncio/events.py [hunk 3]
```diff
@@ -101,7 +101,8 @@ class TimerHandle(Handle):
     __slots__ = ['_scheduled', '_when']
 
     def __init__(self, when, callback, args, loop, context=None):
-        assert when is not None
+        if when is None:
+            raise ValueError('when cannot be None')
```
Comment (AlexWaygood) [inline Lib/asyncio/events.py]: Would a `TypeError` be better here, since `NoneType` is a distinct type?

**Comment by kumaraditya303** [comment]
cc @gvanrossum This only changes AssertionError with exceptions, this does not removes all asserts as you suggested on bpo

File: Lib/asyncio/events.py (hunk 3 above)
Comment (kumaraditya303) [reply Lib/asyncio/events.py]: I think ``ValueError`` is better as ``None`` is in value context here.

**Comment by AlexWaygood** [comment]
Apologies, I should have been clearer in my review. My point wasn't that you should use double-quotes instead of single-quotes — my point was that Python's `TypeError` messages usually look like this:

`policy must be an instance of AbstractEventLoopPolicy or None, not 'str'`

Rather than

`policy must be an instance of AbstractEventLoopPolicy or None, not str`

**Comment by kumaraditya303** [comment]
> Apologies, I should have been clearer in my review. My point wasn't that you should use double-quotes instead of single-quotes — my point was that Python's `TypeError` messages usually look like this:
> 
> `policy must be an instance of AbstractEventLoopPolicy or None, not 'str'`
> 
> Rather than
> 
> `policy must be an instance of AbstractEventLoopPolicy or None, not str`

Done

**Comment by AlexWaygood** [comment]
> Done

Thanks!

File: Lib/asyncio/events.py [hunk 4]
```diff
@@ -101,7 +101,8 @@ class TimerHandle(Handle):
     __slots__ = ['_scheduled', '_when']
 
     def __init__(self, when, callback, args, loop, context=None):
-        assert when is not None
+        if when is None:
```
Comment (asvetlov) [inline Lib/asyncio/events.py]: TimerHandle is never instantiated explicitly; the check should be moved into loop.call_later()/loop.call_at() methods

**Review by asvetlov — CHANGES_REQUESTED** [review]

File: Lib/asyncio/events.py (hunk 4 above)
Comment (kumaraditya303) [reply Lib/asyncio/events.py]: In tests, it is tested by instantiating  directly, doing this would require removing the test.

Comment (asvetlov) [reply Lib/asyncio/events.py]: I think, removing the test and adding new ones for call_later()/call_at() is what we should do.

Comment (kumaraditya303) [reply Lib/asyncio/events.py]: Done

**Comment by kumaraditya303** [comment]
I have made the requested changes; please review again

File: Lib/asyncio/base_events.py [hunk 5]
```diff
@@ -717,6 +717,8 @@ def call_at(self, when, callback, *args, context=None):
 
         Absolute time corresponds to the event loop's time() method.
         """
+        if when is None:
+            raise ValueError("when cannot be None")
```
Comment (gvanrossum) [inline Lib/asyncio/base_events.py]: Arguably a TypeError?

File: Lib/asyncio/events.py [hunk 6]
```diff
@@ -101,7 +101,6 @@ class TimerHandle(Handle):
     __slots__ = ['_scheduled', '_when']
 
     def __init__(self, when, callback, args, loop, context=None):
-        assert when is not None
```
Comment (gvanrossum) [inline Lib/asyncio/events.py:104]: Why did you delete this check?

File: Lib/asyncio/base_events.py (hunk 5 above)
Comment (asvetlov) [reply Lib/asyncio/base_events.py]: Agree with @gvanrossum 
`call_later()` should be updated as well.
Now it throws `TypeError: unsupported operand type(s) for +: 'float' and 'NoneType'`
The error text is very unclear.

File: Lib/asyncio/events.py (hunk 6 above)
Comment (asvetlov) [reply Lib/asyncio/events.py:104]: `loop.call_at(...)` is the only place where TimeHandler created, call_at() has a check now.

Do you prefer to check twice?

Comment (gvanrossum) [reply Lib/asyncio/events.py:104]: Is TimeHandle not a public API?

Comment (asvetlov) [reply Lib/asyncio/events.py:104]: Good question, I have no strong answer for it.
From my perspective, Handle, TimerHandle, Future, Task, and transports are part of public API but their constructors are not. All these objects should be instantiated by factory methods only.
Also, all these constructors are not documented.

Currently, asyncio checks `coro` argument in `Task.__init__` but blindly trusts all arguments in transport constructors.
I believe, the task case is a result of a very old and discouraged design when we used to allow `task = Task(async_func())` usage. Now it is soft-deprecated without raising a warning.

Regarding `TimerHandle`: the direct `th = asyncio.TimerHandle(...)` call doesn't make any sense; there is no public way to register it in the loop. An alone handle is simply useless.

File: Lib/asyncio/base_events.py (hunk 5 above)
Comment (kumaraditya303) [reply Lib/asyncio/base_events.py]: What should be updated in `call_later`?

File: Lib/asyncio/events.py (hunk 6 above)
Comment (gvanrossum) [reply Lib/asyncio/events.py:104]: Okay.

File: Lib/asyncio/base_events.py (hunk 5 above)
Comment (asvetlov) [reply Lib/asyncio/base_events.py]: Please implement a check similar to added to `call_at()`:
```
if delay is None:
    raise TypeError(...)
```

Comment (kumaraditya303) [reply Lib/asyncio/base_events.py]: Done

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks!

**Comment by gvanrossum** [comment]
Congrats Kumar, and thanks Andrew for guiding Kumar towards success here!

File: Lib/asyncio/base_events.py [hunk 7]
```diff
@@ -706,6 +706,8 @@ def call_later(self, delay, callback, *args, context=None):
         Any positional arguments after the callback will be passed to
         the callback when it is called.
         """
+        if delay is None:
+            raise TypeError('delay must not be None')
```
Comment (serhiy-storchaka) [inline Lib/asyncio/base_events.py:710]: It is redundant. A TypeError will be raised at the next line, by `self.time() + delay`.

File: Lib/asyncio/base_events.py [hunk 8]
```diff
@@ -717,6 +719,8 @@ def call_at(self, when, callback, *args, context=None):
 
         Absolute time corresponds to the event loop's time() method.
         """
+        if when is None:
+            raise TypeError("when cannot be None")
```
Comment (serhiy-storchaka) [inline Lib/asyncio/base_events.py:723]: Why test only for None? What about other invalid types?

File: Lib/asyncio/base_events.py (hunk 7 above)
Comment (asvetlov) [reply Lib/asyncio/base_events.py:710]: My thought when I suggested the change was: better to explicitly warn about `None` argument than a little cryptic `unsupported operand type(s) for +: 'NoneType' and 'int'`.
I believe not only exception type but also a proper message can help users to avoid silly mistakes.

File: Lib/asyncio/base_events.py (hunk 8 above)
Comment (asvetlov) [reply Lib/asyncio/base_events.py:723]: The same as above: provide descriptive error message for obvious mistakes, raise default errors for all other combinations.

File: Lib/asyncio/base_events.py (hunk 7 above)
Comment (serhiy-storchaka) [reply Lib/asyncio/base_events.py:710]: Why is `None` special for this argument? I do not know any precedents of checkinjg only for None if other types can be invalid too. It looks like an antipattern to me. We either check for allowed types, or do not check either and let an error be raised later when perform an unsupported operation.
